spec/prometheus/client/counter_spec.rb

```diff
@@ -122,4 +117,57 @@
       end
     end
   end
+
+  describe '#with_labels' do
+    let(:expected_labels) { [:foo] }
+
+    it 'pre-sets labels for observations' do
+      expect { counter.increment }
+        .to raise_error(Prometheus::Client::LabelSetValidator::InvalidLabelSetError)
+      expect { counter.with_labels(foo: 'label').increment }.not_to raise_error
+    end
+
+    it 'registers `with_labels` observations in the original metric store' do
+      counter.increment(labels: { foo: 'value1'})
+      counter_with_labels = counter.with_labels({ foo: 'value2'})
+      counter_with_labels.increment(by: 2)
+
+      expect(counter_with_labels.values).to eql({foo: 'value1'} => 1.0, {foo: 'value2'} => 2.0)
+      expect(counter.values).to eql({foo: 'value1'} => 1.0, {foo: 'value2'} => 2.0)
+    end
+
+    context 'when using DirectFileStore' do
+      before do
+        Dir.glob('/tmp/prometheus_test/*').each { |file| File.delete(file) }
+        Prometheus::Client.config.data_store = Prometheus::Client::DataStores::DirectFileStore.new(dir: '/tmp/prometheus_test')
+      end
+
+      let(:expected_labels) { [:foo, :bar] }
+      
+      it "doesn't corrupt the data files" do
+        counter_with_labels = counter.with_labels({ foo: 'longervalue'})
+
+        # Initialize / read the files for both views of the metric
```

Comment:
The steps in this test feel very particularly chosen, but it's not clear to a reader why. Would you be up for writing up why these operations specifically would cause corruption in a comment?

Reply:
No. I prefer people to work it out for themselves.

Reply:
Just kidding,that's a good idea, i'll do that tomorrow :)

Reply:
Added a 3rd commit with the explanation. It's *long*, sorry, I don't know how to properly explain it shorter without the explanation being more cryptic than the code.